md-proto/WebAssembly.md

```diff
@@ -0,0 +1,2037 @@
+instead treated as if the section was absent.
+
+The Names Section contains an [array] of function name descriptors, which each
+describe names for the function with the corresponding index in the module, and
+which contain:
+ - The function name, a [string].
+ - The names of the locals in the function, an [array] of [strings].
+
+> Name data is represented as an explicit section in the binary format, however
+in the text format it may be represented as an integrated part of the syntax for
+functions rather than as a discrete section.
+
+> The expectation is that, when a binary WebAssembly module is viewed in a
+browser or other development environment, the names in this section will be used
+as the names of functions and locals in the [text format](TextFormat.md).
+
+### Unknown Sections
+
+Modules may also contain unknown sections, with names other than those
+corresponding to known sections above. They have no semantic effect.
+
+### Function Bodies
+
+Function bodies contain an [array] of [types], which are declarations for
+locals, and a sequence of instructions.
+
+#### Positions Within A Function Body
+
+A *position* within a function refers to an element of the instruction sequence.
+
+> In the binary encoding, positions are represented as byte offsets; in the text
+format, positions are represented with a special syntax.
+
+
+Validation
+--------------------------------------------------------------------------------
+
+### Validation Type
+
+A *validation type* is either a [type](#types), `any`, which satisfies any type
+check, or `void`, which satisfies no type check.
+
+> Validation types are only used during validation, and not during execution.
+
+> `any` is effectively serving as a kind of bottom type here, and `void` is
+effectively serving as a kind of top type here. Note that in some type theory
+contexts, it's common to use these names for exactly opposite purposes.
+
+#### Validation Type Merge
+
+To merge two validation types:
+ - If either validation type is `any`, the result is the other validation type.
+ - Otherwise, if the validation types are the same, the result is that
+   validation type.
+ - Otherwise, validation fails.
+
+### Module Validation
+
+TODO: Describe module validation.
+
+### Function Validation
+
+For the duration of the validation of a function body, several data structures
+are created:
+ - A *control-flow stack*, with entries each containing
+    - A [validation type](#validation-type).
+    - A *limit* value.
+ - A *[type] stack*.
+ - A *current position*.
+
+During function validation, if either stack is empty when an element is to be
+popped from it, or an element is accessed by index less than zero or not less
+than the stack's length, validation fails. If a [validation type] is popped from
+the type stack such that the length of the type stack becomes less than the
+control-flow stack top's limit value, validation fails.
+
```

Comment:
Just a suggestion but 'greater than or equal to the stacks length' reads better to me than 'not less
+than'. It might be obvious to you, but perhaps defining a direction for the stack would help to. Then the spec could consistently use 'less' for a check at one end and 'greater' for a check at the other end.